lib/bulk_insert/worker.rb

```diff
@@ -55,20 +55,35 @@ def add_all(rows)
 
     def save!
       if pending?
-        sql = "INSERT INTO #{@table_name} (#{@column_names}) VALUES "
         @now = Time.now
 
         rows = []
+        table = @table_name.gsub('"', '')
+        primary_key = @connection.primary_key(table)
+        prefetch = @connection.prefetch_primary_key?(table)
+        seq_name = @connection.pk_and_sequence_for(table)[1] if prefetch
+
+        rows = []
         @set.each do |row|
           values = []
           @columns.zip(row) do |column, value|
-            value = @now if value == :__timestamp_placeholder
+            if prefetch && column.name == primary_key
+              value = @connection.next_sequence_value(seq_name)
+            elsif value == :__timestamp_placeholder
+              value = @now
+            end
+
             values << @connection.quote(value, column)
           end
-          rows << "(#{values.join(',')})"
+          rows << values.join(', ')
         end
 
-        sql << rows.join(",")
+        case database_type
```

Comment:
This is a nice encapsulation, but I think I'd like to see it simplified further. Can you wrap this case statement in a method, maybe called something like `construct_insert_sql` or something? Then it can all collapse down to a single line in the `save!` method: `@connection.execute(construct_insert_sql(rows))`.